File: runtime/typescript/packages/core/tests/harness/replay-verifier.test.ts
import {
  ReferenceReplayVerifier,
} from "../../src/index";
import {
  ReplayJournalRecord,
  ReplayVerificationRequest,
} from "../../src/model/index";

describe("ReferenceReplayVerifier", () => {
  it("passes identical normalized replay records", () => {
    const record = new ReplayJournalRecord({
      kind: "turn",
      type: "turn_end",
      turnId: "turn-1",
      iteration: 1,
      status: "success",
    });

    const result = new ReferenceReplayVerifier().verify(
      new ReplayVerificationRequest({ expected: [record], actual: [record] }),
    );

    expect(result.status).toBe("passed");
    expect(result.mismatches).toEqual([]);
    expect(result.expectedCount).toBe(1);
    expect(result.actualCount).toBe(1);
  });

  it("reports mismatches with generated contract types", () => {
    const result = new ReferenceReplayVerifier().verify(
      new ReplayVerificationRequest({
        expected: [new ReplayJournalRecord({ kind: "summary", sessionId: "session-1", status: "success" })],
        actual: [new ReplayJournalRecord({ kind: "summary", sessionId: "session-1", status: "error" })],
      }),
    );

    expect(result.status).toBe("failed");
    expect(result.mismatches?.[0]).toMatchObject({
      index: 0,
      message: "Replay record mismatch",
    });
  });
});

File: runtime/go/prompty/model/replay_verifier_test.go
package prompty

import "testing"

func TestReferenceReplayVerifierPassesIdenticalRecords(t *testing.T) {
	record := ReplayJournalRecord{
		Kind:      ReplayRecordKindTurn,
		Type:      stringPtr("turn_end"),
		TurnId:    stringPtr("turn-1"),
		Iteration: int32Ptr(1),
		Status:    ptrReplayRecordStatus(ReplayRecordStatusSuccess),
	}

	result := ReferenceReplayVerifier{}.Verify(ReplayVerificationRequest{
		Expected: []ReplayJournalRecord{record},
		Actual:   []ReplayJournalRecord{record},
	})

	if result.Status != ReplayVerificationStatusPassed || len(result.Mismatches) != 0 {
		t.Fatalf("unexpected result: %#v", result)
	}
	if result.ExpectedCount != 1 || result.ActualCount != 1 {
		t.Fatalf("unexpected counts: %#v", result)
	}
}

func TestReferenceReplayVerifierReportsMismatches(t *testing.T) {
	result := ReferenceReplayVerifier{}.Verify(ReplayVerificationRequest{
		Expected: []ReplayJournalRecord{{
			Kind:      ReplayRecordKindSummary,
			SessionId: stringPtr("session-1"),
			Status:    ptrReplayRecordStatus(ReplayRecordStatusSuccess),
		}},
		Actual: []ReplayJournalRecord{{
			Kind:      ReplayRecordKindSummary,
			SessionId: stringPtr("session-1"),
			Status:    ptrReplayRecordStatus(ReplayRecordStatusError),
		}},
	})

	if result.Status != ReplayVerificationStatusFailed {
		t.Fatalf("expected failure: %#v", result)
	}
	if result.Mismatches[0].Index != 0 || result.Mismatches[0].Message != "Replay record mismatch" {
		t.Fatalf("unexpected mismatch: %#v", result.Mismatches[0])
	}
}

func ptrReplayRecordStatus(value ReplayRecordStatus) *ReplayRecordStatus {
	return &value
}
